src/main/java/org/tugraz/sysds/runtime/lineage/LineageItemUtils.java

```diff
@@ -436,6 +501,30 @@ private static void rReplace(LineageItem current, LineageItem liOld, LineageItem
 		current.setVisited();
 	}
 	
+	public static void replaceDagLeaves(ExecutionContext ec, LineageItem root, CPOperand[] newLeaves) {
+		LineageItem[] newLIleaves = LineageItemUtils.getLineage(ec, newLeaves);
+		HashMap<String, LineageItem> newLImap = new HashMap<>();
+		for (int i=0; i<newLIleaves.length; i++)
+			newLImap.put(newLeaves[i].getName(), newLIleaves[i]);
+
+		//find and replace the old leaves
+		HashSet<LineageItem> oldLeaves = new HashSet<>();
+		rGetDagLeaves(oldLeaves, root);
+		for (LineageItem leaf : oldLeaves) {
+			if (leaf.getType() != LineageItemType.Literal)
+				replace(root, leaf, newLImap.get(leaf.getName()));
+		}
+	}
+
+	public static void rGetDagLeaves(HashSet<LineageItem> leaves, LineageItem root) {
+		if (root.isLeaf())
+			leaves.add(root);
+		else {
+			for (LineageItem li : root.getInputs())
+				rGetDagLeaves(leaves, li);
```

Comment:
Please, consider memoization here. As this is a DAG, a node might be reachable from multiple parents. In the worst-case (e.g., stacked diamond shapes), not doing memoization could lead to exponential instead of linear search. Given the lack of visited flags, you could simple pass a HashSet with node IDs along, add the ID of processed nodes and return if already processed.